api/src/main/java/uk/gov/dvsa/mot/vehicle/hgv/HgvVehicleProvider.java

```diff
@@ -0,0 +1,120 @@
+
+import javax.ws.rs.client.Client;
+import javax.ws.rs.core.Response;
+
+public class HgvVehicleProvider {
+    private static final Logger logger = Logger.getLogger(HgvVehicleProvider.class);
+
+    private HgvConfiguration configuration;
+
+    @Inject
+    public void setHgvConfiguration(HgvConfiguration configuration) {
+        this.configuration = configuration;
+    }
+
+    public Vehicle getVehicle(String registration) throws Exception {
+
+        Vehicle vehicle;
+        TestHistory[] vehicleTestHistory;
+
+        try {
+            CompletableFuture<Vehicle> vehicleFuture = CompletableFuture.supplyAsync(() -> getHgvVehicle(registration));
+
+            CompletableFuture<TestHistory[]> vehicleTestHistoryFuture = CompletableFuture.supplyAsync(
+                    () -> getHgvVehicleTestHistory(registration));
+
+            vehicle = vehicleFuture.get();
+            vehicleTestHistory = vehicleTestHistoryFuture.get();
+        } catch (ExecutionException | InterruptedException e) {
+            logger.error("HGV calls execution error", e);
+            throw new Exception("HGV calls execution error" + e.getStackTrace());
+        }
+
+        if (vehicle != null) {
+            vehicle.setTestHistory(vehicleTestHistory);
+        }
+
+        return vehicle;
+    }
+
+    private Vehicle getHgvVehicle(String registration) {
+        ResponseVehicle response = getHgvResponse(registration,  "/vehicle/moth", ResponseVehicle.class);
+
+        return response != null ? response.getVehicle() : null;
+    }
+
+    private TestHistory[] getHgvVehicleTestHistory(String registration) {
+        ResponseTestHistory response = getHgvResponse(registration, "/testhistory/moth", ResponseTestHistory.class);
+
+        return response != null ? response.getTestHistory() : null;
+    }
+
+    private <T> T getHgvResponse(String registration, String endpointToCall, Class<T> type) {
+
+        try {
+            logger.trace("Entering getHgvResponse(), endpoint to call: " + endpointToCall);
+
+            Response response = getClient().target(configuration.getApiUrl() + endpointToCall)
+                    .queryParam("identifier", registration)
+                    .request()
+                    .header("x-api-key", configuration.getApiKey())
+                    .get();
+
+            int status = response.getStatus();
+            logger.info("HGV response status code: " + status);
+
+            switch (status) {
+                case 200:
+                    return response.readEntity(type);
+                case 204:
+                    response.close();
+                    return type.newInstance();
+                default:
+                    response.close();
+                    throw new IOException("Invalid http response code");
+            }
+        } catch (IOException e) {
+            logger.error(String.format("IO error during communication with api from HGV endpoint : %s", endpointToCall), e);
```

Comment:
that can be collapsed into one catch clause with pipe operator. One does not need to have exception name in the message while exception is logged anyway

Reply:
fixed